### ISSUE TITLE
Making oem be a superset of vmx

### ISSUE THREAD

**Issue by stuart-mclaren-hpe** [OSInside/kiwi#1425]
## Problem description

This is more a query than a bug -- hopefully that's ok.  (I'd issues signing up for Google Groups when I tried to post to the mailing list). 

First up, I'm new to kiwi. I've been able to build some Centos 7 images superfast -- I'm impressed with what I've seen so far.  But it's not clear how to best create a swap partition on a vmx image.

I've seen some other queries about this, eg

https://groups.google.com/forum/#!searchin/kiwi-images/swap|sort:date/kiwi-images/2YrRhtM5KZ0/Z2DM9Q7tBwAJ

https://groups.google.com/forum/#!searchin/kiwi-images/swap|sort:date/kiwi-images/pr7ryvxjHCw/Bgs-CoxXDwAJ

They seem to suggest some workarounds, eg creating swap on first boot.

I seem to have been able to create a vmx swap partition directly from kiwi, without separate workaround commands.

After reading through the code I tried mixing in some `oemconfig` options into the example Centos 7 config.xml, like this:

```
        <type image="vmx" primary="true" format="vmdk" filesystem="ext3" kernelcmdline="rhgb no_timer_check console=tty0 console=ttyS0,115200 net.ifnames=0 biosdevname=0" bootloader="grub2" firmware="uefi">
            <size unit="G">8</size>
            <oemconfig>
                <oem-swap>true</oem-swap>
                <oem-swapsize>400</oem-swapsize>
                <oem-multipath-scan>false</oem-multipath-scan>
            </oemconfig>
        </type>
```

On the VM I checked that this produced what I wanted: a live 400M swap partition: 

```
# swapon --show
NAME      TYPE      SIZE USED PRIO
/dev/vda3 partition 400M   0B   -2
```

For completeness, the partition table looked like this:

```
# fdisk -l /dev/vda
WARNING: fdisk GPT support is currently new, and therefore in an experimental phase. Use at your own discretion.

Disk /dev/vda: 8589 MB, 8589934592 bytes, 16777216 sectors
Units = sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disk label type: gpt
Disk identifier: EE3D4058-DDBF-4EB9-8B64-B38DD4B6CA92


#         Start          End    Size  Type            Name
 1         2048         6143      2M  BIOS boot       p.legacy
 2         6144        47103     20M  EFI System      p.UEFI
 3        47104       866303    400M  Linux swap      p.swap
 4       866304     16777182    7.6G  Linux filesyste p.lxroot
```

After reading the code I'm wondering if this may be working 'by accident'. As far as I can tell, only the first existing `oemconfig` snippet will be read from the xml file, so if you have an oem type with it's `oemconfig` section I think it's possibly only the first one will ever be read (I haven't tested that this is true).

In short: is using `oemconfig` with `vmx` in this way supported?

## Expected behaviour

From reading the documentation I think I'd have expected this to produce an xml schema error saying that `oemconfig` is not valid when used with type vmx.

Obviously, it's doing what I want though (!).

## OS and Software information
+ KIWI version: latest
+ Operating system: opensuse container
+ OBS version: centos 7

Thanks

**Comment by schaefi** [comment]
> But it's not clear how to best create a swap partition on a vmx image.

The vmx image type is meant to be a simple disk image, mostly used in cloud frameworks. There is no support for partition management in this image type. That's why you don't see the swap setup documented for vmx.

> In short: is using oemconfig with vmx in this way supported?
> After reading the code I'm wondering if this may be working 'by accident'.

You encountered an interesting case. You are also right setting oemconfig with a swap setup is effective for building a vmx image. It's not something we do not recommend people to do but you can expect this to work.

> From reading the documentation I think I'd have expected this to produce an xml schema error > saying that oemconfig is not valid when used with type vmx

yes this is correct we also wanted to add this restriction but then found that some people with older image descriptions from the former kiwi version (legacy kiwi, written in perl) used section combinations like you did which were not planned that way but we also don't wanted to let them
run into a validation error when there is no easy schema translation for their use case.

So given somebody uses vmx in combination with oemconfig we could only auto translate this correctly when we turn vmx into oem. This is probably not what a customer wants or expects. Thus we currently live with the inconsistency but don't document it and also give other advises to bring swap into a vmx setup. The most obvious one is to switch image="vmx" to image="oem" because it does not hurt.

I can understand if the additional resize functionality that comes with an oem could be unwanted and some customers doesn't want this "extra source for issues".

You can live with the current setup, as of know we have not planned to restrict the schema, but it can happen in the future as it would be a valid request when the legacy kiwi version is EOL. There is however no date set when this will happen.

I hope all that makes sense to you ?

My personal advise is to switch to oem if that suits your use case. If this is not what you want I would live with the current setup with a big fat warning in the XML and switch to oem if there will be a schema restriction in the future. 

Let me know what you think

Thanks

**Comment by stuart-mclaren-hpe** [comment]
Wow. What an excellent open and thorough explanation -- thank you Marcus.

I have one follow on query.

Could you explain why 'clear' support for vmx swap partitions isn't considered a requirement? Consider for example, this (again, really good) documentation for azure images:
 
https://osinside.github.io/kiwi/working_with_images/vmx_setup_for_azure.html

Is wanting to add a swap partition to vmx images such as the one above unwarranted for some reason? Tweaking the boot partition seems allowed, so the idea of having some control over vmx swap seems valid (well, to me anyway!).  

Similarly, a use case where you have ephemeral VMs that only boot once (eg as part of a Continuous Integration system) the best solution would seem to be bypassing the resize-on-boot logic, but again allowing a swap partition.

Basically, I'm wondering if explicit support for a vmx swap partition is something that doesn't intrinsically make sense, or is more something that's a judgement call / not implemented due to finite developer bandwidth?

Thanks.

**Comment by schaefi** [comment]
I think the answer to your question is related to what we thought is the use case of a vmx - simple disk image. The idea of vmx is to be used in single-shot machines or in cloud environments.

* The first area runs a machine, starts a task and be done with it. My understanding was that this machine should not swap to perform that task. If it swaps the task should better run on a more powerful machine

* The second area covers cloud environments which are based on a virtualization technology. This means any machine is controlled by some hypervisor, e.g kvm, hyperV, vmware, xen ... you name it. All of them manages memory and storage capacity on that level. swap space as it was designed is meant to kick in if you run out of main memory. The consequence of swap is a significant loss of performance. On virtual machines you control those resources such that a machine should never swap.

Those two are the main reasons for keeping vmx as simple as possible and not care about swap. The reason why we need to care for the boot partition is because there were filesystems the bootloader has no support for, which means it has to read its data from a filesystem the bootloader can read. We had this situation for xfs and also for btrfs in the past. Fortunately nowadays the bootloader can also directly read from it. But this sort of issue can always come back if yet another fancy filesystem reaches the planet and also with care for other archs ;)

Our expectation is if this simple concept of a disk is not enough you switch to the next more complex variant which is an oem image

I'm aware, and you already mentioned it, that there are for sure use cases where you want parts of both concepts. However it was not so easy to draw the line between simple and complex and for the moment this is what we have decided.

I'm happy for the history of this conversation. It will help to adapt if we find out that a swap setup is considered a core disk feature that should be available in any case. At the moment I don't see a high demand to change it though.

For the future we could consider to reach simplicity in the oem image by adding options to switch on/off certain features like swap, disk-resize and so on. This will make vmx an obsolete type and could also be handled by transformation stylesheets for existing descriptions.

We can keep this as an idea for the future.

Thoughts ?

**Comment by stuart-mclaren-hpe** [comment]
Making oem be a superset of vmx (in particular, having a "no-resize" option) and making vmx obsolete seems to make sense. (Though I'm new here, so my opinion shouldn't carry too much weight.) I think it could make it easier for new users to ramp up by removing a possible point of confusion. It would remove the burden of worrying about which kiwi specific abstraction you need to select ("luxury of ignorance").

Thanks again for your detailed response, and a great tool.

**Comment by schaefi** [comment]
> Thanks again for your detailed response, and a great tool.

Thanks we are happy if it is useful to you :)

> Making oem be a superset of vmx 

Actually I think that summarizes the conversation best. I'll change the topic of this issue to that summary and we keep it as a refactoring task. That way nothing gets lost.

Thanks for your feedback and experience report

**Comment by schaefi** [comment]
We discussed in the sprint plan how to continue with the idea of making oem a superset of vmx.
From a coding perspective this is already possible, so it's just a matter of describe it better than today. Here is what we came up with:

* a vmx described as oem would look like this

```xml
<type image="oem" ...>
    <oemconfig>
        <oem-resize>false</oem-resize>
    </oemconfig>
</type>
```

* an existing vmx type needs to be translated like above to an oem through a XSLT style sheet

* the following runtime checks must be more specific. The need to depend on the
   setting for oem-resize and the installiso/installstick attributes

   - check_dracut_module_for_disk_oem_in_package_list
   - check_dracut_module_for_oem_install_in_package_list

We think it makes sense to add this modifications such that there will be only one disk
type in the future.

At the moment other tasks are more important but this one stays in the backlog for later

Thanks
